### PR TITLE
Memoize pattern match in GetAliases

### LINKED ISSUE #102676: Optimize pattern check for aliases for GetAliases API
When finding aliases for GetAliases API, we step through all indices and for each index we step its aliases to see whether they match the requested patterns.

https://github.com/elastic/elasticsearch/blob/33cbb8ee231564fe464e1829e95a4b1ff839f2c0/server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java#L430-L435

When an aliases is assigned to many indices, the pattern matching can be performed multiple times which is not necessary. If number of patterns are large, the problem can be even worse because the pattern matching has a 3rd loop through the patterns for each alias.

https://github.com/elastic/elasticsearch/blob/33cbb8ee231564fe464e1829e95a4b1ff839f2c0/server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java#L436-L440

We should look into ways to avoid the unnecessary pattern matchings.

### PR DESCRIPTION
**Pull request by bcully**
Previously, when GetAliases was called with multiple indices that share an alias, findAliases would recheck the alias against the supplied alias patterns for each index that used it. This change memoizes the pattern check to avoid that redundant work.

Closes #102676

### CODE REVIEW

**Comment by bcully** [comment]
Added a basic test of multiple indices sharing an alias.

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -881,24 +881,30 @@ private void findAliasInfo(final String[] aliases, final String[] possibleMatche
 
         boolean matchAllAliases = patterns.length == 0;
 
+        // memoize pattern match against aliases to avoid repeatedly matching when multiple indices share an alias
+        HashMap<String, Boolean> seenAliases = new HashMap<>();
```
Comment (ankikuma) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:885]: Are we at all worried that this HashMap could get very large ?

**Comment by ankikuma** [comment]
Perhaps we can add a test to testFindAliasWithExclusion() as well just to see if it works for the exclusion case as well.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
Comment (bcully) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:885]: It's not impossible, but the caller of this function is already building and [returning a map](https://github.com/elastic/elasticsearch/blob/57846444cc185f3a9a34c068ed566eb462eb565f/server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java#L811) that is roughly the same size (number of keys up to the sum of aliases across all concrete indices, with keys being pointers to existing strings and values being new boxes). So ~2x transient usage? The entries themselves are pretty small (pointer, boxed bool) so you'd need a lot to matter in absolute terms I think.

Having said that I don't know whether we have any limits or typical upper bounds on the number of aliases in a cluster. I also don't know whether we have any limits on the number of concurrent GetAliases calls that we admit. Either of those could I suppose make the memory usage of this map matter, though my guess is that we'd be in a somewhat pathological environment at this point, where we were having other problems not related to the size of this allocation.

I suppose if we needed to we could bound this thing as some sort of LRU but I would guess that we'd have to then know how to size it, which depends probably on answers to the questions in the paragraph above about total number of aliases and max concurrency.

**Comment by bcully** [comment]
> Perhaps we can add a test to testFindAliasWithExclusion() as well just to see if it works for the exclusion case as well.

Added in https://github.com/elastic/elasticsearch/pull/113901

**Pull request merged**